--- env/linux.py	2007-11-04 00:59:09.000000000 +0300
+++ env/linux.py	2009-02-12 20:45:33.000000000 +0300
@@ -17,7 +17,7 @@
 	checkPkgConfig(env, 'gtkmm-2.4')
 	checkPkgConfig(env, 'libxml++-2.6')
 	checkPkgConfig(env, 'gtkglext-1.0')
-	checkPkgConfig(env, 'libgtksourceviewmm-1.0')
+	checkPkgConfig(env, 'gtksourceviewmm-2.0')
 	if not conf.CheckCXXHeader('osg/Node'):
 		raise Exception('ERROR: You have to install OSG development package before compiling osgedit')
 	return conf.Finish()
