--- CMakeLists.txt	2009-04-17 17:29:39.000000000 +0400
+++ CMakeLists.txt	2009-04-17 17:30:45.000000000 +0400
@@ -42,7 +42,11 @@
 #######################################
 # enable CUDA support
 #######################################
+
+OPTION(ENABLE_CUDA "Enable cuda dor speedup calculations" ON)
+IF (ENABLE_CUDA)
 INCLUDE(${CMAKE_SOURCE_DIR}/CMakeModules/cuda/FindCuda.cmake)
+ENDIF (ENABLE_CUDA)
 
 
 # Make the headers visible to everything
