--- fs/reiser4/plugin/file/file_conversion.c.orig	2009-02-07 10:14:56.642211768 +0000
+++ fs/reiser4/plugin/file/file_conversion.c	2009-02-07 10:18:41.778876844 +0000
@@ -686,7 +686,7 @@
 	start = pos & (PAGE_CACHE_SIZE - 1);
 	end = start + len;
 
-	page = __grab_cache_page(mapping, index);
+	page = grab_cache_page_write_begin(mapping, index, flags);
 	*pagep = page;
 	if (!page)
 		return -ENOMEM;
